### PR TITLE
Support arbitrary local file paths, drop implementation concept.

### PR DESCRIPTION
**Pull request by aomarks**
- Benchmarks are now specified as arbitrary file paths. There are no longer any requirements about file layout. There is no longer a concept of *implementation*. As long as the file is accessible from the server root (`cwd` or `--root` flag), it will be served. You can point to an HTML file, or to a directory. If it's a directory, we'll serve its `index.html`.

  `tach foo.html`
  `tach bar/baz`
  `tach foo.html?i=1 foo.html?i=2`

- You can also alias these paths (same for remote URLs) so that they are easier to read in the result table. By default we just use the server-root-relative path as the label (or the URL in the case of remote URLs):

  `tach baz=bar/baz.html`
 
- NPM dependency swaps now work by walking the filesystem to find the `package.json` nearest to each benchmark file, and mounting a swapped version of that corresponding `node_modules/` on a dedicated server.

Also addressed some review comments from https://github.com/PolymerLabs/tachometer/pull/35 and https://github.com/PolymerLabs/tachometer/pull/36:

- Parallelize NPM install and server startup.
- Simplified server session code, and the logic around waiting for results in manual mode.

### CODE REVIEW

File: src/format.ts [hunk 1]
```diff
@@ -244,20 +244,23 @@ const queryParamsDimension: Dimension = {
...
 
 const browserDimension: Dimension = {
   label: 'Browser',
-  format: (r: ResultStats) =>
-      `${r.result.browser.name}\n${r.result.browser.version}`,
+  format: (r: ResultStats) => {
+    if (r.result.userAgent === '') {
+      // This happens with remote URLs, since we don't get a chance to
+      // instrument any requests to find the user agent.
+      // TODO We could make one no-op request just to probe the user agent.
```
Comment (justinfagnani) [inline src/format.ts]: webdriver can also execute JS to retrieve the UA

File: src/specs.ts [hunk 2]
```diff
@@ -9,20 +9,13 @@
  * rights grant found at http://polymer.github.io/PATENTS.txt
  */
 
-import * as fsExtra from 'fs-extra';
 import * as path from 'path';
-import * as url from 'url';
+import {URL} from 'url';
```
Comment (justinfagnani) [inline src/specs.ts]: URL is also a global, since Node 10, I believe

File: src/specs.ts [hunk 3]
```diff
@@ -38,102 +31,77 @@ export async function specsFromOpts(opts: Opts): Promise<BenchmarkSpec[]> {
...
       throw new Error(
-          `Browser ${b} is not yet supported, ` +
+          `Browser ${b} is not supported, ` +
           `only ${[...validBrowsers].join(', ')} are currently supported`);
     }
   }
 
-  const remoteUrls = [];
-  const localNames: {name: string, queryString: string}[] = [];
-  let anyLocalNamesAreStar = false;
```
Comment (justinfagnani) [inline src/specs.ts]: what's the plan to allow automatic running of a bunch of benchmarks?

File: src/specs.ts [hunk 4]
```diff
@@ -144,14 +112,24 @@ export async function specsFromOpts(opts: Opts): Promise<BenchmarkSpec[]> {
     if (a.name !== b.name) {
       return a.name.localeCompare(b.name);
     }
-    if (a.queryString !== b.queryString) {
-      return a.queryString.localeCompare(b.queryString);
+    if (a.url.kind !== b.url.kind) {
```
Comment (justinfagnani) [inline src/specs.ts]: If this if for the comparison table, I would actually prefer that the specs are in order that they came on the command line.

File: src/specs.ts (hunk 2 above)
Comment (aomarks) [reply src/specs.ts]: Not available even after adding `@types/node`.

File: src/specs.ts (hunk 3 above)
Comment (aomarks) [reply src/specs.ts]: Use bash wildcards.

File: src/specs.ts (hunk 4 above)
Comment (aomarks) [reply src/specs.ts]: Done.

File: README.md [hunk 5]
```diff
@@ -338,9 +286,10 @@ will expire.
 ## Usage
 
 ```sh
-tach *
-tach my-bench-a my-bench-b
-tach http://example.com/a http://example.com/b
+tach foo.html
```
[the remark this answers is not in the document]
Comment (aomarks) [reply README.md]: Done.

File: src/format.ts (hunk 1 above)
Comment (aomarks) [reply src/format.ts]: Huh cool, I didn't know https://developer.mozilla.org/en-US/docs/Web/API/NavigatorID/userAgent existed. Will do later.

**Pull request merged**